### PR TITLE
Remove momenttype and use reduction initialization infrastructure

### LINKED ISSUE #22901: Base.std, Base.var, Base.quantile fail on Array{Any,1}
`Base.std`, `Base.var`, `Base.quantile` fail with an input of type `Array{Any,1}` containing numerical values. This behavior is different from the other statistic functions in `Base`.
```Julia
a = [1,2,3,4, "A", "B"]
b = [5,6,7,8, "A", "B"]
typeof(a)  # Array{Any,1}

# error
Base.std(a[1:4])           # -> ERROR: MethodError: no method matching zero(::Type{Any})
Base.var(a[1:4])           # -> ERROR: MethodError: no method matching zero(::Type{Any})
Base.quantile(a[1:4], 0.5) # -> ERROR: MethodError: Cannot `convert` an object
                           #    of type Type{Any} to an object of type AbstractFloat

# ok
Base.mean(a[1:4])
Base.median(a[1:4])
Base.middle(a[1:4])
Base.stdm(a[1:4], 2.5)
Base.varm(a[1:4], 2.5)
Base.cor(a[1:4], b[1:4])
Base.cov(a[1:4], b[1:4])

```

My version info:
```
Julia Version 0.6.0
Commit 9036443 (2017-06-19 13:05 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)
```

### PR DESCRIPTION
**Pull request by andreasnoack**
for determining the return type.

Adjust return type determination in `_quantile`

Fixes #22901

### CODE REVIEW

**Comment by tkelman** [comment]
This is used in DataArrays and a few other packages and would be safer to deprecate

File: base/deprecated.jl [hunk 1]
```diff
@@ -1569,6 +1569,10 @@ end
 
 @deprecate String(io::GenericIOBuffer) String(take!(copy(io)))
 
+@eval Base begin
```
Comment (tkelman) [inline base/deprecated.jl]: this file is in Base? can set the extra boolean argument to false so it's not exported

File: base/deprecated.jl [hunk 2]
```diff
@@ -1591,6 +1591,10 @@ end
 @deprecate readstring(filename::AbstractString) read(filename, String)
 @deprecate readstring(cmd::AbstractCmd) read(cmd, String)
 
+@eval Base begin
```
Comment (tkelman) [inline base/deprecated.jl]: this eval shouldn't be needed

**Comment by iamed2** [comment]
Would this fix https://github.com/JuliaLang/julia/issues/21786 ?

**Comment by andreasnoack** [comment]
It appears so
```julia
julia> A = fill(1KiloGram, 2, 2)
2×2 Array{SIUnits.SIQuantity{Int64,0,1,0,0,0,0,0,0,0},2}:
 1 kg   1 kg
 1 kg   1 kg

julia> mean(A, 1)
1×2 Array{SIUnits.SIQuantity{Float64,0,1,0,0,0,0,0,0,0},2}:
 1.0 kg   1.0 kg
```

**Pull request merged**